<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Reproducible Workflows with R and Sequencing Data</title>
    <meta charset="utf-8" />
    <meta name="author" content="Adrian Sven Geissler" />
    <script src="libs/header-attrs/header-attrs.js"></script>
    <link href="libs/remark-css/default.css" rel="stylesheet" />
    <link href="libs/remark-css/hygge.css" rel="stylesheet" />
    <link href="libs/remark-css/metropolis.css" rel="stylesheet" />
    <link href="libs/remark-css/metropolis-fonts.css" rel="stylesheet" />
    <link href="libs/xaringanExtra-extra-styles/xaringanExtra-extra-styles.css" rel="stylesheet" />
    <link href="libs/animate.css/animate.xaringan.css" rel="stylesheet" />
    <link href="libs/tile-view/tile-view.css" rel="stylesheet" />
    <script src="libs/tile-view/tile-view.js"></script>
    <script src="libs/htmlwidgets/htmlwidgets.js"></script>
    <script src="libs/d3/d3.min.js"></script>
    <script src="libs/dagre/dagre-d3.min.js"></script>
    <link href="libs/mermaid/dist/mermaid.css" rel="stylesheet" />
    <script src="libs/mermaid/dist/mermaid.slim.min.js"></script>
    <link href="libs/DiagrammeR-styles/styles.css" rel="stylesheet" />
    <script src="libs/chromatography/chromatography.js"></script>
    <script src="libs/DiagrammeR-binding/DiagrammeR.js"></script>
    <link rel="stylesheet" href="custom.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

.title[
# Reproducible Workflows with R and Sequencing Data
]
.subtitle[
## A Beginner‚Äôs Guide to transcriptomic analysis
]
.author[
### Adrian Sven Geissler
]
.date[
### 7<sup>th</sup> of February 2024
]

---







class: inverse, clear, animated, fadeIn
background-image: url('illustrations/diverse.jpg')
background-position: top
background-size: cover

.slogan[*Diverse* target audience]

.footnote[Photo by Amy Elting on Unsplash]






---
class: inverse
# Presentation takeaways


- .Large[Define "*transcriptomic*" sequencing and its relevance]

--

- .Large[Recognize structure of sequencing data]

--

- .Large[Outline a simple analysis pipeline with bash]

--

- .Large[Identify concept and advantages of a Snakemake workflow]&lt;br/&gt;
  (How I avoid using bash as much as possible)

--

- .Large[Connect a workflow with R analysis scripts]

--

- .Large[Introduction to software management across platforms]

--

- .Large[Fun with üì¶ packages and R tweaks]








---
class: inverse, center, middle

.Large[
Slides:

https://github.com/asgeissler/2024-CopenhagenR-Seminar

Snakemake tutorial workflow:

https://github.com/asgeissler/2024-Snakemake-Intro
]





---

class: inverse, clear, animated, pulse
background-image: url('illustrations/galaxy.jpeg')
background-size: cover

.slogan[
Big data with&lt;br/&gt;
*astronomically* large datasets
]

.footnote[Photo by Greg Rakozy on Unsplash]





---
class: clear, animated, swing
background-image: url('illustrations/meme-big-data.jpeg')
background-size: cover






---
# Big Data: Astronomical or Genomical?

&lt;br /&gt;
&lt;br /&gt;

&lt;img src="illustrations/big-data-comparison.png" width="3000" /&gt;

.footnote[Stephens, et al. PLOS Biology 2015]





---
# Gene expression: From genome, over RNA, to proteins

.center[
&lt;img src="illustrations/dogma.png" width="80%" /&gt;
]


---
class: clear, inverse
background-image: url('illustrations/plans.jpg')
background-size: cover

.footnote[Photo by Evgeniy Surzhan on Unsplash]

--

.slogan[-ome suffix: *A totality of some sort*]




---
# Differences in gene expression drive phenotypes

.pull-left[.Large[

- ‚öôÔ∏è Cell tissue

- üß¨ Genome

- üå≤ Environment &amp; üíä Exposure

- ‚ò†Ô∏è Disease associate factors

]]

.pull-right[
&lt;img src="illustrations/gtex.png" width="80%" /&gt;
]

.footnote[Figure from the GTEx Consortium, Science 2020]








---
class: clear, inverse
background-image: url('illustrations/plans.jpg')
background-size: cover

.slogan[Transcriptomic analysis:&lt;br/&gt; What is happening in the cells?]

.footnote[Photo by Evgeniy Surzhan on Unsplash]








---
class: inverse, clear
background-image: url('illustrations/radio.jpg')
background-size: cover


.footnote[Photo by Dave Weatherall on Unsplash]






---
class: inverse, clear
background-image: url('illustrations/tools.jpg')
background-size: cover


.footnote[Photo by Barn Images on Unsplash]






---
class: inverse
background-image: url('illustrations/shotgun.jpg')
background-size: cover

# Biologist's preferred tool


.footnote[Photo by Stephen Baker on Unsplash]

--
.slogan[Bioinformatics simplified:&lt;br/&gt; Putting the pieces back together...]










---
class: inverse
background-image: url('illustrations/sand-glass.jpeg')
background-size: cover

# The *pre*-R fun bottleneck


.pull-right[
Biological *high-throughput* measurements

- RNA Sequencing data
- Terabytes of `\(&gt; 100\)` compressed txt's
- Specialized tools (bash scripts) 
- Long running times

]

.footnote[Photo by Nathan Dumlao on Unsplash]

--

.pull-right[
Intermediate result:  
Count Matrix `\(&lt; 100\)` MB  
Rows: Genes, Columns: Samples/Patients
]

--

.pull-right[
Data exploration and analysis    
(R scripts / notebooks)

]







---
# Transcriptomic "airway" dataset from 2014

.center[
&lt;img src="illustrations/airway.jpg" width="70%" /&gt;
]

.Large[
- Freely and openly available (without ethics board)

- RNA sequencing datasets for Asthma drug treatment vs control

- `\(4\)` human cell lines (genotypes)
]











---
# Downloading the human genome


.pull-left[
.center[
&lt;img src="illustrations/gencode.png" width="90%" /&gt;
]

.footnote[https://www.gencodegenes.org/]
]

.pull-right[
Copy and paste links from website:

.small[
```
$ URL="https://ftp.ebi.ac.uk/pub/\
  databases/gencode/\
  Gencode_human/release_44/"

$ wget $URL/GRCh38.primary_assembly.genome.fa.gz \
  -O genome.fna.gz
  
$ wget $URL/gencode.v44.basic.annotation.gff3.gz \
  -O genome.gff.gz
```
]
]





---
# üß¨ üï∫ üßç‚Äç‚ôÄÔ∏èüë∂

`\(806\)` MBytes compressed text; `\(3\)` GBytes uncompressed

.large[
```
$ gunzip -c genome.fna.gz| less
&gt;chr1 1
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
...
TGCTCATGAAGTGTGAGTTAATGCACTCAAATCAATGGTTGTGCACGGTTTATATGAATA
TTAGTGATTACAAAATATTATCAATAGACCTTGTCACAACTGTTATTGAAGAACTAATCA
TCTATTGCTTATTTAGGTCTTTCTCTCCTGCCAGAATGTGCGCTCCAGGTGGAGAGGTAT
GTTGCCTTATCCGTGGCTGGATATATAGAGATTCCCACACTGCCTTGCACACGAGCACTG
CTGGGTAAATATTTGTTGGCTGCAGGAAAACGTGAAGGAATAGGCCCTCCAATGGGAGGA
...
&gt;chr2 2
...
```
]





---
# Gene annotations

`\(34\)` MBytes compressed, `\(902\)` MBytes uncompressed

.small[
```
$ gunzip -c genome.gff.gz | less
##gff-version 3
#description: evidence-based annotation of the human genome (GRCh38), version 44 (Ensembl 110)
[...]
chr1	HAVANA	gene	11869	14409	.	+	.	ID=ENSG00000290825.1;[...];gene_name=DDX11L2;[...]
...
```
]

Chromosome 1 has a gene at positions `\(11,869\)` to `\(14,409\)` on the `\(+1\)` strand.

--


&lt;img src="illustrations/browser-gene-close.png" width="90%" /&gt;


Figure from https://genome-euro.ucsc.edu




---
# Dataset access: Sequence Read Archive

.Large[
From publication text: *"The RNA-Seq data is available at..."*
]

.center[
&lt;img src="illustrations/sra-website.png" width="60%" /&gt;
]

https://trace.ncbi.nlm.nih.gov/Traces/?view=study&amp;acc=SRP033351







---
# Downloaded sample files

.large[
- **Option 1**: Using the custom `SRA Toolkit` (bash...)

- **Option 2**: Clicking on website
]

```
$ ls
raw-data    genome.fna.gz   genome.gff.gz

$ ls -1 raw-data
SRR1039508_1.fastq.gz   # (control sample)
SRR1039508_2.fastq.gz
SRR1039509_1.fastq.gz   # (asthma treated case)
SRR1039509_2.fastq.gz
...
```

In this dataset, the files are `\(3.3-7.3\)` GBytes large  
( `\(0.9-1.7\)` GBytes compressed)


---
# Paired-end sequencing data

.pull-left[
.center[
&lt;img src="illustrations/mirror.jpg" width="60%" /&gt;
]

]

.pull-right[

.large[
```
&gt; ls -1 raw-data
SRR1039508_1.fastq.gz
SRR1039508_2.fastq.gz
...
```
]

.Large[
The data contains the "*front*" and the "*back*"
of the sequenced read
]

.footnote[Photo by Taylor Smith on Unsplash]
]





---
# Lots of compressed text files

.large[

```bash
&gt; gunzip -c data/SRR1039508_1.fastq.gz | head
@SRR1039508.1 HWI-ST177:290:C0TECACXX:1:1101:1225:2130 length=63
CATTGCTGATACCAANNNNNNNNGCATTCCTCAAGGTCTTCCTCCTTCCCTTACGGAATTACA
+SRR1039508.1 HWI-ST177:290:C0TECACXX:1:1101:1225:2130 length=63
HJJJJJJJJJJJJJJ########00?GHIJJJJJJJIJJJJJJJJJJJJJJJJJHHHFFFFFD
...
```
]

--

.Large[
That look like one of those **ACTG** genome thingies!
]

&lt;img src="illustrations/cletus.webp" width="15%" /&gt;



---
# Goal: Number of reads mapping to a gene

&lt;img src="illustrations/browser-mapping.png" width="95%" /&gt;





---
class: clear, animated, rotateIn
background-image: url('illustrations/text-lookup.jpg')
background-size: contain







---
# Sequencing quality filtering is needed

.Large[
`\(-10 \cdot log_{10}(\)` Probability of sequencing error `\()\)`

Encoded as single, *printable ASCII* character
]

(details depend on the sequencing platform manufacturer)

&lt;img src="illustrations/phred.png" width="90%" /&gt;

.footnote[https://usegalaxy.be/training-material//topics/metagenomics/tutorials/metatranscriptomics/slides-plain.html]







---
# Efficient sequence mapping algorithm is needed

.large[
- Short reads: `\(&lt; 100\)` nucleotides (characters)

- This dataset: `\(13-34\)` million read pairs, per sample

- Human genome: `\(&gt;  3.1\)` billion base pairs (characters)
]

--

.Large[
~~Naive string lookup~~
]

&lt;img src="illustrations/cletus.webp" width="15%" /&gt;








---
# Efficient sequence mapping algorithms

.wide[
.Large[
2. Search via a phonebook-like index structure  
  (*e.g.*, Burrows‚ÄìWheeler Transformation)
3. *Approximate* via sequence spectrum  
  (*k*-mer frequency)
]

Suggested reading:&lt;br /&gt;
Li, Durbin. Bioinformatics 2009 https://doi.org/10.1093/bioinformatics/btp324&lt;br/&gt;
Patro, *et al.* Nature Methods 2017 https://doi.org/10.1038/nmeth.4197

]

.narrow[
&lt;img src="illustrations/phone.jpg" width="100%" /&gt;

.small[Photo by Johnny Briggs on Unsplash]
]








---
# Outline processing pipeline

.center[
<div id="htmlwidget-d3f48bfd1288b6f7d23c" style="width:100%;height:100%;" class="DiagrammeR html-widget"></div>
<script type="application/json" data-for="htmlwidget-d3f48bfd1288b6f7d23c">{"x":{"diagram":"\ngraph TB\n    gff[Gene annotations]\n    fna[Genomic sequences]\n    ix[Index]\n    raw[Sequencing data]\n    clean[Quality Control]\n    map[Genome mapping]\n    x[Gene expression matrix]\n    fun[R analysis]\n    raw --> clean\n    clean --> map\n    fna --> ix\n    ix --> map\n    map --> x\n    gff --> x\n    x --> fun\n"},"evals":[],"jsHooks":[]}</script>
]





---
# Bash programming primer

```
qcTool raw-data/sample1.fastq.gz &gt; clean-data/sample1.fastq.gz
qcTool raw-data/sample2.fastq.gz &gt; clean-data/sample2.fastq.gz
qcTool raw-data/sample3.fastq.gz &gt; clean-data/sample3.fastq.gz
qcTool raw-data/sample4.fastq.gz &gt; clean-data/sample4.fastq.gz
...
```

--

Slightly better solution

```
for i in raw-data/*fastq.gz ; do
  filename=$(basename $i)
  qcTool $i &gt; clean-data/$filename
done
```




---
# Conceptual bash pipeline

.pull-left[
<div id="htmlwidget-83a1902f3731c37de917" style="width:100%;height:100%;" class="DiagrammeR html-widget"></div>
<script type="application/json" data-for="htmlwidget-83a1902f3731c37de917">{"x":{"diagram":"\ngraph TB\n    gff[Gene annotations]\n    fna[Genomic sequences]\n    ix[Index]\n    raw[Sequencing data]\n    clean[Quality Control]\n    map[Genome mapping]\n    x[Gene expression matrix]\n    fun[R analysis]\n    raw --> clean\n    clean --> map\n    fna --> ix\n    ix --> map\n    map --> x\n    gff --> x\n    x --> fun\n"},"evals":[],"jsHooks":[]}</script>
]


.pull-right[
.small[A file `concept-pipeline.bash`]

```
for i in raw-data/*.fastq.gz ; do
  filename=$(basename $i)
  qcTool $i &gt; clean-data/$filename
done

makeIndex genome.fna.gz index

for i in clean-data/*.fastq.gz ; do
  filename=$(basename $i .fastq.gz)
  mapTool index $i &gt; mapping/$filename.bam
done

countExpression annotation.gff.gz mapping/*  \
  &gt; matrix.txt
```
]




---
# https://github.com/asgeissler/2024-Snakemake-Intro/

.center[
&lt;img src="illustrations/terminal.png" width="70%" /&gt;
]




---
class: inverse
# Bash concerns

.Large[

- Adding new samples? Re-run pipeline &amp; over-write files?

- Running steps in parallel?

- Consistency when changing parameters?

- Software management?
]




---
# Snakemake to the rescue üêçü¶∏‚Äç‚ôÇÔ∏è

.center[
&lt;img src="illustrations/snakemake-intro.png" width="65%" /&gt;
]

.small[Slide by Johannes K√∂ster
https://slides.com/johanneskoester/snakemake-tutorial
]



---
# Towards Snakemake üë®‚Äçüç≥ "cooking" rules üìö üç≥ 


.large[
```
$ cat Snakefile

rule make_my_index:
    input:
        'genome.fna.gz'
    output:
        'index'
    shell:
        """
        makeIndex {input} {output}
        """
```
]

--

`snakemake --cores all make_my_index`

or

`snakemake --cores all index`






---
# Wildcard placeholders

```
rule qc_cleaning:
    input:
        'raw-data/{sample}.fastq.gz'
    output:
        'clean-data/{sample}.fastq.gz'
    shell:
        """
        qcTool {input} &gt; {output}
        """
```

--

```
$ snakemake --cores all clean-data/sample1.fastq.gz
Select jobs to execute...
rule qc_cleaning:
    input: raw-data/sample1.fastq.gz
    output: clean-data/sample1.fastq.gz
    jobid: 1
    reason: Missing output files: clean-data/sample1.fastq.gz
    wildcards: sample=sample1
```




---
# Rules with list and named parameters

.small[
```
xs = ['sample1', 'sample2', 'sample3', 'sample4']

rule count_expression:
    input:
        genome = 'genome.gff.gz',
        mapping = expand('mapping/{sample}.bam', sample = xs)
    output:
        'matrix.txt'
    shell:
        """
        echo "I am the named genome input: {input.genome}"
        echo "We are the input mappings: {input.mapping}"
        echo "I am the first mapping: {input.mapping[0]}"
        
        countExpression {input} &gt; {output}
        """

```
]

--

```
$ snakemake --cores all matrix.txt
...
I am the named genome input: genome.gff.gz
We are the input mappings: mapping/sample1.bam mapping/sample2.bam mapping/sample3.bam...
I am the first mapping: mapping/sample1.bam
```




---
class: inverse, clear


.Large[
```
$ qcTool
bash: qcTool: command not found
```
]


.Large[
What are the actual tools, and how to use them?
]

--

.Large[
Thank you for this short intro to bash...
]

--

.Large[
But: Can we do with less bash coding, please?
]








---
# üêç Snakemake üéÅ  wrappers

.center[
&lt;img src="illustrations/snakemake-wrapper.png" width="70%" /&gt;
]

.small[https://snakemake-wrappers.readthedocs.io/en/stable/wrappers/star/index.html]



---
# üì¶ Automatic dependencies resolution

&lt;img src="illustrations/snakemake-wrapper2.png" width="90%" /&gt;

.small[https://snakemake-wrappers.readthedocs.io/en/stable/wrappers/star/index.html]






---
# Tutorial workflow

&lt;img src="illustrations/tutorial-workflow.jpg" width="80%" /&gt;



---
#Tutorial workflow


.large[
```
$ git clone https://github.com/asgeissler/2024-Snakemake-Intro/
$ cd 3-Tutorial-Workflow
$ snakemake --cores all --use-conda --conda-frontend=mamba \
     gene-expression-matrix.tsv
```
]

Should run in `\(&lt; 15\)` min on a notebook

```
$ head -n 3 gene-expression-matrix.tsv
Geneid            Chr  Start End         Strand Length SRR1039508 SRR1039509 SRR1039512 SRR1039513
ENSG00000223587.2 chr3 11745 24849          +   13105  0           1         0          0
ENSG00000224918.1 chr3 53348 54346          -   999    0           0         0          0

$  grep ENSG00000163884.4 gene-expression-matrix.tsv
ENSG00000163884.4  chr3 126342635 126357408 - 14774    220         3384      128        1518
```






---
class: inverse
background-image: url('illustrations/sand-glass.jpeg')
background-size: cover

# The end of the *pre*-R fun bottleneck

.pull-right[
Intermediate result:  
Count Matrix `\(&lt; 100\)` MB  
Rows: Genes, Columns: Samples/Patients
]

--

.pull-right[
&lt;img src="illustrations/Rlogo.png" width="80%" /&gt;
]





---
# Data and software from Bioconductor


.clear[
.thirds[
&lt;img src="illustrations/logo_bioconductor.gif" width="100%" /&gt;

https://bioconductor.org/
]
.double-r[.Large[
"The mission of the Bioconductor project is
[...] **open source** software
[...] rigorous and **reproducible** analysis
[...] welcoming **community** of developers and data scientists."
]]

]

--

.clear[
.wide[.large[
*"The Bioconductor community is a long-term user of __S4__ and has produced much
of the best material about its __effective use__. "*&lt;br/&gt;
‚Äî Hadley Wickham
]]
.narrow[
&lt;img src="illustrations/cover-advancedR.png" width="60%" /&gt;
]
]





---
# {airway} üì¶ 


```r
install.packages("BiocManager")
BiocManager::install("airway")
```


```r
library(airway)
data(airway)
airway
```

```
## class: RangedSummarizedExperiment 
## dim: 63677 8 
## metadata(1): ''
## assays(1): counts
## rownames(63677): ENSG00000000003 ENSG00000000005 ... ENSG00000273492
##   ENSG00000273493
## rowData names(10): gene_id gene_name ... seq_coord_system symbol
## colnames(8): SRR1039508 SRR1039509 ... SRR1039520 SRR1039521
## colData names(9): SampleName cell ... Sample BioSample
```




---
# {airway} üì¶ 

Memo R pipes: `g(f(x))` is `x |&gt; f() |&gt; g()`&lt;br/&gt;
(`|&gt;` ligature, typed `| &gt;`)



```r
airway |&gt;
  assay() |&gt;
  as.matrix() |&gt;
  _[1:5, 1:5]
```

```
##                 SRR1039508 SRR1039509 SRR1039512 SRR1039513 SRR1039516
## ENSG00000000003        679        448        873        408       1138
## ENSG00000000005          0          0          0          0          0
## ENSG00000000419        467        515        621        365        587
## ENSG00000000457        260        211        263        164        245
## ENSG00000000460         60         55         40         35         78
```


.Large[
The full
matrix of gene (row) expression levels per sample (columns) from
the *pre-R* fun step.
]





---
# Sample meta information

 
`colData`: Column meta information data


```r
airway |&gt; colData() |&gt; as_tibble() |&gt; head(n = 3) |&gt; kable()
```



|SampleName |cell    |dex   |albut |Run        | avgLength|Experiment |Sample    |BioSample    |
|:----------|:-------|:-----|:-----|:----------|---------:|:----------|:---------|:------------|
|GSM1275862 |N61311  |untrt |untrt |SRR1039508 |       126|SRX384345  |SRS508568 |SAMN02422669 |
|GSM1275863 |N61311  |trt   |untrt |SRR1039509 |       126|SRX384346  |SRS508567 |SAMN02422675 |
|GSM1275866 |N052611 |untrt |untrt |SRR1039512 |       126|SRX384349  |SRS508571 |SAMN02422678 |

.Large[Asthma drug treated (`trt`) or control (`untrt`)]

.footnote[
`kable()` from {knitr} üì¶ makes pretty tables in these slides&lt;br/&gt;
S4 class: `colData(airway)` accessed the object's slot `airway@colData`
]






---
# üß¨ üôã‚Äç‚ôÇÔ∏è Our Genes


```r
airway |&gt;
  rowData() |&gt; as_tibble() |&gt;
  select(gene_id, gene_name, chromosome = seq_name,
         gene_seq_start, gene_seq_end, seq_strand) |&gt;
  filter(chromosome == 'X') |&gt;
  arrange(gene_seq_start) |&gt;
  mutate_if(is.numeric, prettyNum, big.mark = ',') |&gt;
  head(n = 3) |&gt; kable()
```



|gene_id         |gene_name       |chromosome |gene_seq_start |gene_seq_end |seq_strand |
|:---------------|:---------------|:----------|:--------------|:------------|:----------|
|ENSG00000228572 |LL0YNC03-29C1.1 |X          |170,410        |171,758      |1          |
|ENSG00000182378 |PLCXD1          |X          |192,989        |220,023      |1          |
|ENSG00000178605 |GTPBP6          |X          |220,025        |230,886      |-1         |

.large[tibbles are the better dataframes {tidyverse} üì¶ üßπ]





---
# *Tidy* data exploration

.pull-left[

```r
# One  dexamethasone drug treated library
my.data &lt;-
  airway |&gt;
  assay() |&gt;
  as_tibble(rownames = 'gene_id') |&gt;
  select(gene_id, expression = SRR1039509)
my.data |&gt; head(n = 3) |&gt; kable()
```



|gene_id         | expression|
|:---------------|----------:|
|ENSG00000000003 |        448|
|ENSG00000000005 |          0|
|ENSG00000000419 |        515|

]

.pull-right[

```r
my.genes &lt;-
  airway |&gt;
  rowData() |&gt;
  as_tibble() |&gt;
  select(gene_id, gene_name, chromosome = seq_name,
         gene_seq_start, gene_seq_end, seq_strand)
*my.genes |&gt; str()
```

```
## tibble [63,677 √ó 6] (S3: tbl_df/tbl/data.frame)
##  $ gene_id       : chr [1:63677] "ENSG00000000003" "ENSG00000000005" "ENSG00000000419" "ENSG00000000457" ...
##  $ gene_name     : chr [1:63677] "TSPAN6" "TNMD" "DPM1" "SCYL3" ...
##  $ chromosome    : chr [1:63677] "X" "X" "20" "1" ...
##  $ gene_seq_start: int [1:63677] 99883667 99839799 49551404 169818772 169631245 27938575 196621008 143815948 53362139 41040684 ...
##  $ gene_seq_end  : int [1:63677] 99894988 99854882 49575092 169863408 169823221 27961788 196716634 143832827 53481768 41067715 ...
##  $ seq_strand    : int [1:63677] -1 1 -1 -1 1 -1 1 -1 -1 1 ...
```
]





---
# Use the join!

.pull-left[
.Large[

```r
my.genes.data &lt;- full_join(
  my.genes,
  my.data,
  by = 'gene_id'
)
```
]
]

.pull-right[
&lt;img src="illustrations/meme-join.jpg" width="70%" /&gt;

]


|gene_id         |gene_name |chromosome | gene_seq_start| gene_seq_end| seq_strand| expression|
|:---------------|:---------|:----------|--------------:|------------:|----------:|----------:|
|ENSG00000000003 |TSPAN6    |X          |       99883667|     99894988|         -1|        448|
|ENSG00000000005 |TNMD      |X          |       99839799|     99854882|          1|          0|
|ENSG00000000419 |DPM1      |20         |       49551404|     49575092|         -1|        515|







---
# A first genomic expression figure with {ggplot2}


.pull-left[
.large[

```r
plot1 &lt;-
  my.genes.data |&gt;
  filter(chromosome == '1') |&gt;
  mutate_at('seq_strand',
            as.character) |&gt;
  ggplot(aes(
    x = gene_seq_start, 
    y = expression,
    color = seq_strand),
  ) +
  geom_point(alpha = .5) +
  scale_y_log10() +
  xlab('Chromosome 1 [bp]')
```
]
]

--

.pull-right[
![](seminar-R-snakemake-workflow_files/figure-html/unnamed-chunk-36-1.png)&lt;!-- --&gt;

&lt;img src="illustrations/chromosome.jpg" width="60%" /&gt;
]







---
# Motivation normalization

.pull-left[
.large[

```r
scatter.plot &lt;-
  airway |&gt;
  assay() |&gt;
  as_tibble() |&gt;
  ggplot(aes(x = SRR1039517,
             y = SRR1039513)) +
  geom_point(alpha = 0.5) +
  geom_abline(slope = 1,
              color = 'blue') +
  scale_x_log10() +
  scale_y_log10()
```
]
]

.pull-right[
![](seminar-R-snakemake-workflow_files/figure-html/unnamed-chunk-39-1.png)&lt;!-- --&gt;
]







---
# Which waterway has more fish?

.pull-left[
&lt;img src="illustrations/fish1.jpg" width="100%" /&gt;
]
.pull-right[
&lt;img src="illustrations/fish2.jpg" width="100%" /&gt;
]


.footnote[
Photos by Mega Caesaria and  Rick Wallace on Unsplash
]








---
# RNA-seq data normalization



.large[
Potential method, size-factor normalization:

Per library (column) compute the
median ratio of gene counts relative to geometric mean per gene (row)
]


Suggested reading:

Anders S and Huber W.
[Differential expression analysis for sequence count data.](https://www.nature.com/articles/npre.2010.4282.1)
Nature Preceding 2010


RNA-seq analysis workflow tutorial on Bioconductor:

https://bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html








---
# Differential expression analysisw with {DESeq2}


.large[

```r
library(DESeq2)

des &lt;-
  airway |&gt;
  # convert to DataSet for DESeq
  # analyse expression by cell line and dex treatment
  DESeqDataSet(design = ~ cell + dex) |&gt;
  # Run DESeq2 normalization and regression models
  DESeq()

# Statistical results
res &lt;- results(des, tidy = TRUE)
```
]


RNA-seq analysis workflow tutorial on Bioconductor:

https://bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html





---
# Statistical analysis of expression changes


$$
\text{Expression Fold Change} =  \frac
{\text{Expression in condition A}}
{\text{Expression in condition B}}
$$

--



|   A|   B| Fold Change| log2|
|---:|---:|-----------:|----:|
| 100| 100|         1.0|    0|
| 100|  50|         2.0|    1|
| 100| 200|         0.5|   -1|


--

.large[
Testing for statistical significance and false discovery rate (FDR) adjustment
]



---
# Analysis results

.pull-left[

```r
res |&gt;
  ggplot(aes(log2FoldChange, - log10(padj))) +
  geom_point() +
  theme_bw(18)

ggsave('volcano.jpeg')
```
]
.pull-right[
&lt;img src="volcano.jpeg" width="2400" /&gt;
]






---
# Hello academic paper writing phase


```r
my.analysis &lt;-
  res |&gt;
  filter(padj &lt;= 0.05, abs(log2FoldChange) &gt;= 1) |&gt;
  left_join(airway |&gt; rowData() |&gt; as_tibble(), c('row' = 'gene_id')) |&gt;
  select(gene_id = row, gene_name, log2FoldChange, FDR = padj) |&gt;
  arrange(FDR)
```


|gene_id         |gene_name | log2FoldChange| FDR|
|:---------------|:---------|--------------:|---:|
|ENSG00000152583 |SPARCL1   |      -4.57|   &lt; 1e-100|
|ENSG00000165995 |CACNB2    |      -3.29|   &lt; 1e-100|
|ENSG00000120129 |DUSP1     |      -2.94|   &lt; 1e-100|
|ENSG00000101347 |SAMHD1    |      -3.76|   &lt; 1e-100|
|ENSG00000189221 |MAOA      |      -3.35|   &lt; 1e-100|
|ENSG00000211445 |GPX3      |      -3.73|   &lt; 1e-100|





---
# Dealing with the office devil {openxlsx}

.large[

```r
library(openxlsx)

my.analysis |&gt;
  write.xlsx('my-analysis.xlsx')
```
]

.center[
&lt;img src="excel.png" width="60%" /&gt;
]





---
class: inverse, clear
background-image: url('illustrations/shotgun.jpg')
background-size: cover

.slogan[Happy customer!]


.footnote[Photo by Stephen Baker on Unsplash]






---
# üè¥‚Äç‚ò†Ô∏è R and üêç


.pull-left[
```
rule my_analysis:
    input:
        script = 'analysis.R',
        data = 'gene-expression-matrix.tsv'
    output:
      'volcano.jpeg',
      'my-analysis.xlsx'
    shell:
        "RScript {input.script}"
```
]
.pull-right[
.large[
- Pro tip: Define script as input (triggers rerun on changes)

- Explicitly list data input, such that script gets executed at the right time

- Increase reproducibility by declaring software environment with conda
]

]

.footnote[
Snakemake Handbook:

https://snakemake.readthedocs.io/en/stable/snakefiles/deployment.html#integrated-package-management
]





---
class: clear
background-image: url('illustrations/shoes.jpg')
background-position: bottom
background-size: cover

.slogan[
There is no "one size fits all" expression analysis workflow
]

.footnote[Photo by Julian Hochgesang on Unsplash]





---
class: inverse
# Presentation takeaways


- .Large[Define "*transcriptomic*" sequencing and its relevance]

- .Large[Recognize structure of sequencing data]

- .Large[Outline a simple analysis pipeline with bash]

- .Large[Identify concept and advantages of a Snakemake workflow]&lt;br/&gt;
  (How I avoid using bash as much as possible)

- .Large[Connect a workflow with R analysis scripts]

- .Large[Introduction to software management across platforms]

- .Large[Fun with üì¶ packages and R tweaks]





---
class: inverse, clear
background-image: url('illustrations/copenhagen.jpg')
background-position: top
background-size: cover

.slogan[
Thank you for joining!
]

.footnote[Photo by Peter Lloyd on Unsplash]


---
class: inverse, clear
background-image: url('illustrations/copenhagen.jpg')
background-position: top
background-size: cover

.slogan[
Supervisors: SE Seemann, J Gorodkin

Independent Research Fund Denmark
]

.footnote[Photo by Peter Lloyd on Unsplash]

    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"ratio": "16:9",
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
// add `data-at-shortcutkeys` attribute to <body> to resolve conflicts with JAWS
// screen reader (see PR #262)
(function(d) {
  let res = {};
  d.querySelectorAll('.remark-help-content table tr').forEach(tr => {
    const t = tr.querySelector('td:nth-child(2)').innerText;
    tr.querySelectorAll('td:first-child .key').forEach(key => {
      const k = key.innerText;
      if (/^[a-z]$/.test(k)) res[k] = t;  // must be a single letter (key)
    });
  });
  d.body.setAttribute('data-at-shortcutkeys', JSON.stringify(res));
})(document);
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
